utils/SimpleJson.cpp

```diff
@@ -0,0 +1,522 @@
+    if (node.getType() != Type::None)
+    {
+#if SIMPLE_JSON_CACHE_DIAG
+        printf("\n Found in cache for path %s\n", path);
+#endif
+        return node;
+    }
+    char cachedPath[1024];
+    cachedPath[0] = 0;
+    return findInternal(path, cachedPath, 0, _cache);
+}
+
+SimpleJson SimpleJson::findInternal(const char* const path,
+    char* const cachedPath,
+    size_t cachedPathSize,
+    JsonPathCache& cache)
+{
+    auto token = StringTokenizer::tokenize(path, strlen(path), '.');
+    if (token.empty() || Type::Object != _type)
+    {
+        return SimpleJsonNone;
+    }
+
+    if (cachedPathSize)
+    {
+        strcat(cachedPath, ".");
+        cachedPathSize++;
+    }
+    strncpy(cachedPath + cachedPathSize, token.start, token.length);
+    cachedPathSize += token.length;
+    cachedPath[cachedPathSize] = 0;
+
+    auto property = findInCache(cachedPath, cachedPathSize, cache);
+    if (property.getType() == Type::None)
+    {
+        property = findProperty(_cursorIn + 1, token.start, token.length);
+        cache.put(cachedPath, cachedPathSize, property._cursorIn, property._cursorOut);
+#if SIMPLE_JSON_CACHE_DIAG
+        printf("\n Put in cache path %s\n", cachedPath);
+#endif
+    }
+#if SIMPLE_JSON_CACHE_DIAG
+    else
+    {
+        printf("\n Found in cache for path %s\n", cachedPath);
+    }
+#endif
+
+    if (property.getType() == SimpleJson::Type::None || !token.next)
+    {
+        return property;
+    }
+
+    return (property.getType() == SimpleJson::Type::Object)
+        ? property.findInternal(token.next, cachedPath, cachedPathSize, cache)
+        : SimpleJsonNone;
+}
+
+Optional<int64_t> SimpleJson::getIntValue() const
+{
+    if (Type::Integer != _type)
+    {
+        return Optional<int64_t>();
+    }
+    char _buffer[33];
+    strncpy(_buffer, _cursorIn, size());
+    int64_t out;
+    return (1 == sscanf(_buffer, "%" SCNd64, &out)) ? Optional<int64_t>(out) : Optional<int64_t>();
+}
+
+Optional<double> SimpleJson::getFloatValue() const
+{
+    if (Type::Float != _type)
+    {
+        return Optional<double>();
+    }
+    char _buffer[33];
+    strncpy(_buffer, _cursorIn, size());
+    double out;
+    return (1 == sscanf(_buffer, "%lf", &out)) ? Optional<double>(out) : Optional<double>();
+}
+
+bool SimpleJson::getStringValue(const char*& out, size_t& outLen) const
+{
+    if (Type::String != _type || size() < 2)
+    {
+        return false;
+    }
+    out = _cursorIn + 1;
+    outLen = size() - 2;
+    return true;
+}
+
+Optional<bool> SimpleJson::getBoolValue() const
+{
+    if (Type::Boolean != _type)
+    {
+        return Optional<bool>();
+    }
+    if (size() == 4 && !strncmp(_cursorIn, "true", 4))
+    {
+        return Optional<bool>(true);
+    }
+    if (size() == 5 && !strncmp(_cursorIn, "false", 5))
+    {
+        return Optional<bool>(false);
+    }
+    return Optional<bool>();
+}
+
+Optional<SimpleJsonArray> SimpleJson::getArrayValue() const
```

Comment:
I think we should do lazy evaluation. Array does not need internal data storage. The iterators can just find the next lement position when we do ++.